bftengine/src/bftengine/KeyStore.cpp

```diff
@@ -0,0 +1,169 @@
+
+void ReplicaKeyStore::serializeDataMembers(std::ostream& outStream) const {
+  serialize(outStream, numOfKeysLimit_);
+  serialize(outStream, keys_);
+}
+
+void ReplicaKeyStore::deserializeDataMembers(std::istream& inStream) {
+  deserialize(inStream, numOfKeysLimit_);
+  deserialize(inStream, keys_);
+}
+
+ReplicaKeyStore ReplicaKeyStore::deserializeReplicaKeyStore(const char* serializedRepStore, const int& size) {
+  std::stringstream ss;
+  ReplicaKeyStore ks;
+  ss.write(serializedRepStore, std::streamsize(size));
+  deserialize(ss, ks);
+  return ks;
+}
+
+bool ReplicaKeyStore::rotate(const uint64_t& chknum) {
+  if (keys_.size() < 2) return false;
+  auto chekPointSeqNum = chknum * seqNumsPerChkPoint_;
+  // Since it's FIFO we need to check the second element only.
+  uint16_t seqNumsSinceKeyExchangeMsg = chekPointSeqNum - keys_[1].seqnum;
+  ConcordAssertGE(seqNumsSinceKeyExchangeMsg, 0);
+  // if checkpoint is less than desired
+  if (seqNumsSinceKeyExchangeMsg < (checkPointsForRotation_ - 1) * seqNumsPerChkPoint_) {
+    return false;
+  }
+  // if somehow rotation wasn't performed on desired checkpoint.
+  ConcordAssertLT(seqNumsSinceKeyExchangeMsg, checkPointsForRotation_ * seqNumsPerChkPoint_);
+  LOG_DEBUG(GL,
+            "KEY EXCHANGE:: Key rotation for replica " << keys_[1].msg.repID << " recieved on seqnum "
+                                                       << keys_[1].seqnum << " rotated on " << chekPointSeqNum);
+  keys_.pop_front();
+  return true;
+}
+
+////////////////////////REPLCIA KEY///////////////////////////
+
+const std::string ReplicaKeyStore::ReplicaKey::getVersion() const { return "1"; }
+
+void ReplicaKeyStore::ReplicaKey::serializeDataMembers(std::ostream& outStream) const {
+  serialize(outStream, msg);
+  serialize(outStream, seqnum);
+}
+
+void ReplicaKeyStore::ReplicaKey::deserializeDataMembers(std::istream& inStream) {
+  deserialize(inStream, msg);
+  deserialize(inStream, seqnum);
+}
+
+ReplicaKeyStore::ReplicaKey::ReplicaKey(const KeyExchangeMsg& other, const uint64_t& seqnum)
+    : msg(other), seqnum(seqnum) {}
+
+//////////////////////CLUSTER KEY STORE////////////////////////
+
+ClsuterKeyStore::ClsuterKeyStore(const uint32_t& clusterSize) : clusterKeys_(clusterSize) {}
+
+bool ClsuterKeyStore::push(const KeyExchangeMsg& kem,
+                           const uint64_t& sn,
+                           const std::vector<IKeyExchanger*>& registryToExchange) {
+  if (kem.repID >= clusterKeys_.size()) {
+    LOG_ERROR(GL, "KEY EXCHANGE: replica id is out of range " << kem.repID);
+    return false;
+  }
+
+  if (!clusterKeys_[kem.repID].push(kem, sn)) return false;
+
+  for (auto ike : registryToExchange) {
+    ike->onNewKey(clusterKeys_[kem.repID].current().msg);
+  }
+  return true;
+}
+
+bool ClsuterKeyStore::rotate(const uint64_t& chknum, const std::vector<IKeyExchanger*>& registryToExchange) {
+  bool ret{};
+  auto idx = 0;
+  for (auto& rks : clusterKeys_) {
```

Comment:
Please give more meaningful name to `rks`.

Reply:
yep, you right